File: Android/app/src/main/java/com/example/supplementtracker/service/CommercialEntitlements.kt
package com.example.supplementtracker.service

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

enum class CommercialPlan {
    FREE,
    PRO,
    COACH
}

enum class CommercialFeature {
    BASIC_TRACKING,
    REMINDERS,
    RECENT_HISTORY,
    ADVANCED_CYCLES,
    UNLIMITED_HISTORY,
    ADHERENCE_ANALYTICS,
    ENCRYPTED_CLOUD_SYNC,
    DATA_EXPORT,
    MULTI_CLIENT,
    COACH_REPORTS
}

enum class BillingPeriod {
    MONTHLY,
    ANNUAL
}

data class CommercialProduct(
    val productId: String,
    val plan: CommercialPlan,
    val billingPeriod: BillingPeriod
)

data class EntitlementSnapshot(
    val plan: CommercialPlan,
    val activeProductId: String? = null
) {
    companion object {
        val Free = EntitlementSnapshot(CommercialPlan.FREE)
    }
}

object CommercialProductCatalog {
    const val PRO_MONTHLY = "oak_pro_monthly"
    const val PRO_ANNUAL = "oak_pro_annual"
    const val COACH_MONTHLY = "oak_coach_monthly"
    const val COACH_ANNUAL = "oak_coach_annual"

    val products = listOf(
        CommercialProduct(PRO_MONTHLY, CommercialPlan.PRO, BillingPeriod.MONTHLY),
        CommercialProduct(PRO_ANNUAL, CommercialPlan.PRO, BillingPeriod.ANNUAL),
        CommercialProduct(COACH_MONTHLY, CommercialPlan.COACH, BillingPeriod.MONTHLY),
        CommercialProduct(COACH_ANNUAL, CommercialPlan.COACH, BillingPeriod.ANNUAL)
    )
}

object EntitlementPolicy {
    private val freeFeatures = setOf(
        CommercialFeature.BASIC_TRACKING,
        CommercialFeature.REMINDERS,
        CommercialFeature.RECENT_HISTORY
    )
    private val proFeatures = freeFeatures + setOf(
        CommercialFeature.ADVANCED_CYCLES,
        CommercialFeature.UNLIMITED_HISTORY,
        CommercialFeature.ADHERENCE_ANALYTICS,
        CommercialFeature.ENCRYPTED_CLOUD_SYNC,
        CommercialFeature.DATA_EXPORT
    )

    fun allows(plan: CommercialPlan, feature: CommercialFeature): Boolean {
        return when (plan) {
            CommercialPlan.FREE -> feature in freeFeatures
            CommercialPlan.PRO -> feature in proFeatures
            CommercialPlan.COACH -> true
        }
    }

    fun maxClients(plan: CommercialPlan): Int? {
        return when (plan) {
            CommercialPlan.FREE, CommercialPlan.PRO -> 1
            CommercialPlan.COACH -> null
        }
    }
}

class EntitlementManager(
    initialSnapshot: EntitlementSnapshot = EntitlementSnapshot.Free
) {
    private val mutableSnapshot = MutableStateFlow(initialSnapshot)
    val snapshot: StateFlow<EntitlementSnapshot> = mutableSnapshot.asStateFlow()

    fun replaceFromStore(snapshot: EntitlementSnapshot) {
        mutableSnapshot.value = snapshot
    }

    fun resetToFree() {
        mutableSnapshot.value = EntitlementSnapshot.Free
    }
}

File: iOS/Services/CommercialEntitlements.swift
import Foundation
import Observation

public enum CommercialPlan: String, CaseIterable, Equatable, Hashable, Sendable {
    case free
    case pro
    case coach
}

public enum CommercialFeature: CaseIterable, Hashable, Sendable {
    case basicTracking
    case reminders
    case recentHistory
    case advancedCycles
    case unlimitedHistory
    case adherenceAnalytics
    case encryptedCloudSync
    case dataExport
    case multiClient
    case coachReports
}

public enum BillingPeriod: Equatable, Sendable {
    case monthly
    case annual
}

public struct CommercialProduct: Equatable, Sendable {
    public let productId: String
    public let plan: CommercialPlan
    public let billingPeriod: BillingPeriod
}

public struct EntitlementSnapshot: Equatable, Sendable {
    public let plan: CommercialPlan
    public let activeProductId: String?

    public init(plan: CommercialPlan, activeProductId: String? = nil) {
        self.plan = plan
        self.activeProductId = activeProductId
    }

    public static let free = EntitlementSnapshot(plan: .free)
}

public enum CommercialProductCatalog {
    public static let proMonthly = "oak_pro_monthly"
    public static let proAnnual = "oak_pro_annual"
    public static let coachMonthly = "oak_coach_monthly"
    public static let coachAnnual = "oak_coach_annual"

    public static let products = [
        CommercialProduct(productId: proMonthly, plan: .pro, billingPeriod: .monthly),
        CommercialProduct(productId: proAnnual, plan: .pro, billingPeriod: .annual),
        CommercialProduct(productId: coachMonthly, plan: .coach, billingPeriod: .monthly),
        CommercialProduct(productId: coachAnnual, plan: .coach, billingPeriod: .annual)
    ]
}

public enum CommercialEntitlementResolver {
    public static func resolve(productIds: some Sequence<String>) -> EntitlementSnapshot {
        let ids = Set(productIds)
        let matches = CommercialProductCatalog.products.filter { ids.contains($0.productId) }
        guard let highest = matches.max(by: { planRank($0.plan) < planRank($1.plan) }) else {
            return .free
        }
        return EntitlementSnapshot(plan: highest.plan, activeProductId: highest.productId)
    }

    private static func planRank(_ plan: CommercialPlan) -> Int {
        switch plan {
        case .free: return 0
        case .pro: return 1
        case .coach: return 2
        }
    }
}

public enum EntitlementPolicy {
    private static let freeFeatures: Set<CommercialFeature> = [
        .basicTracking,
        .reminders,
        .recentHistory
    ]
    private static let proFeatures: Set<CommercialFeature> = freeFeatures.union([
        .advancedCycles,
        .unlimitedHistory,
        .adherenceAnalytics,
        .encryptedCloudSync,
        .dataExport
    ])

    public static func allows(plan: CommercialPlan, feature: CommercialFeature) -> Bool {
        switch plan {
        case .free: return freeFeatures.contains(feature)
        case .pro: return proFeatures.contains(feature)
        case .coach: return true
        }
    }

    public static func maxClients(plan: CommercialPlan) -> Int? {
        switch plan {
        case .free, .pro: return 1
        case .coach: return nil
        }
    }
}

@MainActor
@Observable
public final class EntitlementManager {
    public private(set) var snapshot: EntitlementSnapshot

    public init(initialSnapshot: EntitlementSnapshot = .free) {
        snapshot = initialSnapshot
    }

    public func replaceFromStore(_ snapshot: EntitlementSnapshot) {
        self.snapshot = snapshot
    }

    public func resetToFree() {
        snapshot = .free
    }
}
